source/common/router/config_impl.cc

```diff
@@ -69,29 +69,71 @@ ShadowPolicyImpl::ShadowPolicyImpl(const envoy::api::v2::RouteAction& config) {
   runtime_key_ = config.request_mirror_policy().runtime_key();
 }
 
+class HeaderHashImpl : public HashPolicyImpl::HashImpl {
+public:
+  HeaderHashImpl(const std::string& header_name) : header_name_(header_name) {}
+
+  ~HeaderHashImpl() override {}
+
+  Optional<uint64_t> evaluate(const std::string&, const Http::HeaderMap& headers) const override {
+    Optional<uint64_t> hash;
+
+    const Http::HeaderEntry* header = headers.get(header_name_);
+    if (header) {
+      hash.value(HashUtil::xxHash64(header->value().c_str()));
+    }
+    return hash;
+  }
+
+private:
+  Http::LowerCaseString header_name_;
+};
+
+class IpHashImpl : public HashPolicyImpl::HashImpl {
+public:
+  IpHashImpl() {}
+
+  ~IpHashImpl() override {}
+
+  Optional<uint64_t> evaluate(const std::string& downstream_addr,
+                              const Http::HeaderMap&) const override {
+    Optional<uint64_t> hash;
+    if (!downstream_addr.empty()) {
+      hash.value(HashUtil::xxHash64(downstream_addr));
+    }
+    return hash;
+  }
+};
+
 HashPolicyImpl::HashPolicyImpl(
-    const Protobuf::RepeatedPtrField<envoy::api::v2::RouteAction::HashPolicy>& hash_policy)
-    : header_name_(hash_policy[0].header().header_name()),
-      hash_ip_(hash_policy[0].connection_properties().source_ip()) {
-  // TODO(htuch): Add support for multiple hash policies, #1422.
-  ASSERT(hash_policy.size() == 1);
+    const Protobuf::RepeatedPtrField<envoy::api::v2::RouteAction::HashPolicy>& hash_policies) {
   // TODO(htuch): Add support for cookie hash policies, #1295
-  ASSERT(hash_policy[0].policy_specifier_case() ==
-             envoy::api::v2::RouteAction::HashPolicy::kHeader ||
-         hash_policy[0].policy_specifier_case() ==
-             envoy::api::v2::RouteAction::HashPolicy::kConnectionProperties);
+  hash_impls_.reserve(hash_policies.size());
+
+  for (auto& hash_policy : hash_policies) {
+    switch (hash_policy.policy_specifier_case()) {
+    case envoy::api::v2::RouteAction::HashPolicy::kHeader:
+      hash_impls_.emplace_back(new HeaderHashImpl(hash_policy.header().header_name()));
+      break;
+    case envoy::api::v2::RouteAction::HashPolicy::kConnectionProperties:
+      if (hash_policy.connection_properties().source_ip()) {
+        hash_impls_.emplace_back(new IpHashImpl());
+      }
+      break;
+    default:
+      ASSERT(false);
```

Comment:
Might consider moving to throwing an `EnvoyException` here. I was a bit lazy initially as I expected to fill in all cases (or have constraints to cover them), but if we're writing new code, we should either add a `TODO` about using constraints to catch this/implement or throw.